### PR TITLE
fix: added fractional offset support

### PR DESCRIPTION
**Pull request by rharshit82**
## Description
This is a fix for Issue - https://github.com/kelektiv/node-cron/issues/568
There was a stale pull request in the past but since it was not active, made a fresh PR with support for negative as well as positive offset with minute support - https://github.com/kelektiv/node-cron/pull/685

## Related Issue

Fixes https://github.com/kelektiv/node-cron/issues/568.

## Motivation and Context

It allows negative as well as positive offset with minute spport

## How Has This Been Tested?

Tested in node 16.16.0, ran all the jet test cases and added 2 test cases where it was needed

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] If my change introduces a breaking change, I have added a `!` after the type/scope in the title (see the Conventional Commits standard).

### CODE REVIEW

File: lib/time.js [hunk 1]
```diff
@@ -152,17 +152,27 @@ function CronTime(luxon) {
 			}
 
 			if (typeof this.utcOffset !== 'undefined') {
-				let offset =
-					this.utcOffset >= 60 || this.utcOffset <= -60
-						? this.utcOffset / 60
-						: this.utcOffset;
+				if (typeof this.utcOffset === 'string') {
```
Comment (sheerlox) [inline lib/time.js:155]: `utcOffset` "should be an integer amount representing the number of minutes offset (like 120 for +2 hours or -90 for -1.5 hours)"

Comment (sheerlox) [reply lib/time.js:155]: That's weird since the code didn't support that before you implemented it :thinking: was the test passing? could you please point to that test?
But indeed, that might be a good thing to support strings as well, @intcreator wdyt?

Comment (rharshit82) [reply lib/time.js:155]: This was the test case which was passing as well since we are using parseInt

`
it('should run a job using cron syntax with string format utcOffset', () => {
			const clock = sinon.useFakeTimers();
			const callback = jest.fn();
			const luxon = require('luxon');
			// Current time
			const t = luxon.DateTime.local();
			// UTC Offset decreased by an hour (string format '(+/-)HH:mm')
			const utcOffset = t.offset - 60;
			let utcOffsetString = utcOffset > 0 ? '+' : '-';
			utcOffsetString += ('0' + Math.floor(Math.abs(utcOffset) / 60)).slice(-2);
			utcOffsetString += ':';
			utcOffsetString += ('0' + (utcOffset % 60)).slice(-2);

			const job = new cron.CronJob(
				t.second + ' ' + t.minute + ' ' + t.hour + ' * * *',
				callback,
				null,
				true,
				null,
				null,
				null,
				utcOffsetString
			);

			// tick 1 sec before an hour
			clock.tick(1000 * 60 * 60 - 1);
			expect(callback).toHaveBeenCalledTimes(0);

			// tick 1 sec
			clock.tick(1);
			clock.restore();
			job.stop();
			expect(callback).toHaveBeenCalledTimes(1);
		});
`

Comment (sheerlox) [reply lib/time.js:155]: right, `parsetInt` just ignores whatever is after the semicolon ... those kinds of implicit behaviors are really hazardous :sweat_smile:

File: lib/time.js [hunk 2]
```diff
@@ -152,17 +152,27 @@ function CronTime(luxon) {
...
+					const parts = this.utcOffset.split(':');
+					const hours = parseInt(parts[0]);
+					const minutes = parseInt(parts[1]);
```
Comment (sheerlox) [inline lib/time.js:158]: maybe use `parseInt(parts[1] ?? '0');` in case the colon and minutes are omitted in the string?

**Comment by sheerlox** [comment]
I've updated the PR title since this PR will introduce a breaking change: previously strings were parsed as minutes, now they would be parsed as "HH:mm" strings.

File: lib/time.js [hunk 3]
```diff
@@ -152,17 +152,27 @@ function CronTime(luxon) {
...
+					else this.utcOffset = hours * 60 + minutes;
+				}
+				let offset = this.utcOffset / 60;
+
 				offset = parseInt(offset);
 
+				const remainderMins = Math.abs(this.utcOffset - offset * 60);
+				const remainderMinsString =
+					remainderMins >= 10 ? remainderMins : '0' + remainderMins;
+
```
Comment (sheerlox) [inline lib/time.js:169]: to improve naming consistency, I suggest renaming a few variables: 
- `offset` => `offsetHours`  
- `remainderMins` => `offsetMins`  
- `remainderMinsString` => `offsetMinsStr`

**Comment by rharshit82** [comment]
Also, wanted to highlight here that previously also string behaviour worked fine, just minute support was not there.
Offset : "5:30" translates to "UTC+5"
Offset: "330" also translates to "UTC+5"

File: tests/cron.test.js [hunk 4]
```diff
@@ -834,6 +867,42 @@ describe('cron', () => {
...
 
+		it('should run a job using cron syntax with string format utcOffset with minute', () => {
+			const clock = sinon.useFakeTimers();
+			const callback = jest.fn();
+			const luxon = require('luxon');
+			// Current time
+			const t = luxon.DateTime.local();
+			// UTC Offset decreased by an 45 minutes (string format '(+/-)HH:mm')
+			const utcOffset = t.offset - 45;
+			let utcOffsetString = utcOffset > 0 ? '+' : '-';
```
Comment (sheerlox) [inline tests/cron.test.js:878]: please use the same lines formatting [as your other test](https://github.com/kelektiv/node-cron/pull/685/files#diff-839e01ecdddc69537435c7ad3d7b1a2ce11a1d41644ada6836cb966c48a38519R803-R811) (also "an" in comment needs to be removed)

**Review by sheerlox — CHANGES_REQUESTED** [review]
left a few comments requesting some changes, otherwise lgtm :+1:

**Comment by sheerlox** [comment]
> Also, wanted to highlight here that previously also string behaviour worked fine, just minute support was not there. Offset : "5:30" translates to "UTC+5" Offset: "330" also translates to "UTC+5"

oh my .. you're absolutely right, ["magic" code right here](https://github.com/sheerlox/node-cron/blob/main/lib/time.js#L155-L159) ..

the kind of hidden behaviors that I was talking about earlier, those can really be a real pain :facepalm:

**Comment by rharshit82** [comment]
Have done the changes and tested it. Let me know if documentation changes are required for string support. 

One thing to note is if we are adding string support, there will be cases where someone might put '330' as string offset where our code will take 330 hours as offset. I am thinking of handling this by using string as minutes in cases where : is not provided. What do you think?

**Comment by sheerlox** [comment]
> Have done the changes and tested it. Let me know if documentation changes are required for string support.

Thanks for your changes, and yes, please update the documentation to reflect your changes.

> One thing to note is if we are adding string support, there will be cases where someone might put '330' as string offset where our code will take 330 hours as offset. I am thinking of handling this by using string as minutes in cases where : is not provided. What do you think?

I think the best for now is to avoid introducing a breaking change in this PR, so we can release it on the v2 while we work on v3. I'll convert the codebase to Typescript soon, and I will be handling that issue at that time, which will then be released on v3 :)

**Comment by sheerlox** [comment]
Well actually there's a breaking change since you removed the "magic" code I referenced earlier (treating `utcOffset` as hours if it's `<-60 || >60` ..

What do you think about reintroducing that behaviour for now, so this PR contains no breaking change, then start working on another PR targeting the `beta` branch to start redesigning that feature?

**Comment by sheerlox** [comment]
I ran the test pipeline for this PR, and there's some codestyle compliance issues.

You can run `npm run lint -- --fix` locally to autofix some of the issues, but there's (at least) one that'll still need to be fixed manually (your console will tell you which one after running the command).

**Comment by rharshit82** [comment]
Thank you for reviewing the PR in such a short time. I will be doing all the changes asked above.

File: lib/time.js (hunk 1 above)
Comment (intcreator) [reply lib/time.js:155]: my opinion is that we explicitly support strings. if there was unintended support for it before we want our changes to be backwards compatible. however we should explicitly parse it in a clear way

File: tests/cron.test.js (hunk 4 above)
Comment (intcreator) [reply tests/cron.test.js:880]: thoughts on building a string with template literals instead of mutating a variable? like
```
const utcOffsetString = `
    ${utcOffset > 0 ? '+' : '-'}
    ${('0' + Math.floor(Math.abs(utcOffset) / 60)).slice(-2)}:
    ${('0' + Math.floor(Math.abs(utcOffset) / 60)).slice(-2)}
`;
 ```

**Comment by intcreator** [comment]
I took a look as well. good conversation and this PR is moving in a good direction. I just made one small suggestion

File: tests/cron.test.js (hunk 4 above)
Comment (sheerlox) [reply tests/cron.test.js:880]: that's indeed more readable

File: lib/time.js [hunk 5]
```diff
@@ -152,17 +152,32 @@ function CronTime(luxon) {
...
+				let offsetHours =
 					this.utcOffset >= 60 || this.utcOffset <= -60
 						? this.utcOffset / 60
 						: this.utcOffset;
-				offset = parseInt(offset);
+
+				offsetHours = parseInt(offsetHours);
+
+				const offsetMins = Math.abs(this.utcOffset - offsetHours * 60);
+				const offsetMinsStr = offsetMins >= 10 ? offsetMins : '0' + offsetMins;
```
Comment (sheerlox) [inline lib/time.js:171]: when defining `offsetHours`  the time unit of `this.utcOffset` (hours or minutes) is subject to a condition, but when defining `offsetMins` (which depends `offsetHours`), that condition isn't present.
This makes the `should run a job using cron syntax with numeric format utcOffset with minute` test fail with the following values:
```
this.utcOffset: 15
offsetHours: 15
offsetMins: 885 [ Math.abs(15−15×60) ]
utcZone: UTC+15:885 => invalid UTC offset
```

Please remember to run `npm run test && npm run lint -- --fix` before you commit :wink: 
*there is also a minor codestyle issue that will be fixed by the above command*

**Review by sheerlox — CHANGES_REQUESTED** [review]

**Comment by sheerlox** [comment]
regarding the new test case `should run a job using cron syntax with number format with minute support`, it is failing on my local setup. My timezone is UTC+1, so `t.offset = 60` in the test. Here's a snapshot of the state in `sendAt` while the test is running:

image

the resulting `utcZone` equals `UTC-20:00`, preventing the test success.

I think the changes in https://github.com/kelektiv/node-cron/pull/685/commits/d509e61b6e56a9e2c8fb9b2219ab951c1d8e28b2 need to be reverted. I'm going to work on that while I'm at it so we can push this through ASAP.

**Comment by sheerlox** [comment]
Found a pretty good solution so the test will run the same way regardless of the local timezone.
@rharshit82 @intcreator could you please review my change (https://github.com/kelektiv/node-cron/pull/685/commits/caf1cda67be366ba06219251363e2f205819ac02)?

**Comment by sheerlox** [comment]
I realized while reviewing the PR that we can specify a string offset in the `timezone` parameter (e.g. `UTC+5:30`), making string support on the `offset` parameter redundant, so I'll remove string support.

**Review by sheerlox — APPROVED** [review]

**Comment by rharshit82** [comment]
We should definitely work on changing the dependence on local time zones for test cases in the future as it is very easy to miss edge cases. I checked your changes, Since only HH is supported in the string utcOffset, the test case does not support +xx:yy where yy is >0 and < 60. My timezone is 5:30 so the test case is not passing on my machine. 
Have solved this in this Pr - https://github.com/rharshit82/node-cron/pull/1/files. Please review this and approve it so that I can merge it otherwise you can also make this change in this PR.

**Comment by sheerlox** [comment]
good catch, I didn't put much effort into that test since it's going to be removed soon. I've reviewed and tested your changes, I'll merge your fork PR now.

**Review by sheerlox — APPROVED** [review]
